### PR TITLE
Yaml 2 json

### LINKED ISSUE #12: [FEATURE] Restructure config file
Some plugins may perform multiple tasks:

- Chrome may open multiple windows
- PyCharm may open multiple projects
- VLC should be able to [play a song](https://www.youtube.com/watch?v=5tXuHYq8lF0) and increase the volume to 100%
- Git should be able to send a pull request and open a merge request right afterword.

In the best scenario, two plugins should be able to interact (Spotify: Play a random song, Slack: Post that song on the #music channel).

Let's talk about the possible structure of the `.comeback` file. 

[GitLab's CI format](https://docs.gitlab.com/ee/ci/yaml/) looks to me like a great example.

### PR DESCRIPTION
**Pull request by agamm**
Thanks in advance bru.
fixes: https://github.com/agamm/comeback/issues/12

### CODE REVIEW

File: comeback/main.py [hunk 1]
```diff
@@ -1,25 +1,18 @@
...
 import pathlib
 import sys
-from types import ModuleType
 from typing import Any, Dict, Optional, Tuple, List
 
 import click
-import yaml
 
-from comeback import config
+from comeback import config, utils
```
Comment (yammesicka) [inline comeback/main.py]: Newline, or use the last import line to import the specific function

File: comeback/recipe_manager.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+from typing import Any, Dict, Optional, List
+from comeback import paths
+from comeback import config
+from comeback import plugin_manager
+from comeback.utils import verbose_echo
+
+RECIPE_FILENAME = ".comeback"
+
+
+def read_recipe_file(recipe_path: pathlib.Path) -> Optional[Dict[str, Any]]:
```
Comment (yammesicka) [inline comeback/recipe_manager.py]: Change name to `read_file`

File: comeback/recipe_manager.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        with open(recipe_path, 'r') as fd:
+            return json.load(fd)
+    except IOError as e:
+        click.echo(f'Could not read file {recipe_path} because {e}')
+    except json.JSONDecodeError as exc:
+        click.echo('JSON Error: ' + str(exc))
+    return None
+
+
+def run_recipe(recipe: List[Dict[str, Any]]) -> None:
```
Comment (yammesicka) [inline comeback/recipe_manager.py]: Change name to `run`

File: comeback/recipe_manager.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                exit()
+
+            verbose_echo(f'Starting {plugin_name}...')
+            verbose_echo(f'\tParams {plugin_params}...')
+
+            plugin = plugin_manager.load_plugin(plugin_name)
+            plugin_manager.call_plugin(plugin, **plugin_params)
+
+
+def create_recipe_file() -> pathlib.Path:
```
Comment (yammesicka) [inline comeback/recipe_manager.py]: `create`

File: comeback/recipe_manager.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    verbose_echo('Creating a blank .comeback configuration file.')
+    path = paths.CURRENT_DIR / RECIPE_FILENAME
+    if path.exists():
+        verbose_echo('.comeback file already exists here. Will only touch it.')
+    path.touch()
+    config.add_comeback_path(path)
+    return path
+
+
+def get_recipe_path() -> pathlib.Path:
```
Comment (yammesicka) [inline comeback/recipe_manager.py]: `get_path`

File: comeback/recipe_manager.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        return cwd_path
+
+    last_comeback_used = config.get_last_comeback()
+
+    verbose_echo('No .comeback file found in the current directory, ' +
+                 f'starting last session found ({last_comeback_used})')
+    return last_comeback_used
+
+
+def load_recipe(recipe_path: Optional[pathlib.Path] = None) -> None:
```
Comment (yammesicka) [inline comeback/recipe_manager.py]: `load`

File: comeback/main.py [hunk 7]
```diff
@@ -1,25 +1,18 @@
...
 from comeback import paths
-from comeback import plugins
+from comeback import plugin_manager
+from comeback import recipe_manager
+from comeback.recipe_manager import RECIPE_FILENAME
```
Comment (yammesicka) [inline comeback/main.py]: Change to `recipe`

File: comeback/main.py [hunk 8]
```diff
@@ -1,25 +1,18 @@
...
 
 import click
-import yaml
 
-from comeback import config
+from comeback import config, utils
 from comeback import paths
-from comeback import plugins
+from comeback import plugin_manager
+from comeback import recipe_manager
```
Comment (yammesicka) [inline comeback/main.py]: Don't import twice

File: README.md [hunk 9]
```diff
@@ -21,11 +21,19 @@ If it takes you an average of 2 minutes to open all of your project's tools, the
...
 1) In the project directory, we have `.comeback` recipe:
-```yaml
-vscode: 
-  cwd: ~/dev/myproject
-chrome:
-  url: http://localhost:8080/
+```json
+[  
+   {  
+      "vscode":{
```
Comment (yammesicka) [inline README.md]: Space

File: README.md [hunk 10]
```diff
@@ -21,11 +21,19 @@ If it takes you an average of 2 minutes to open all of your project's tools, the
...
-```yaml
-vscode: 
-  cwd: ~/dev/myproject
-chrome:
-  url: http://localhost:8080/
+```json
+[  
+   {  
+      "vscode":{  
+         "cwd":"~/dev/myproject"
```
Comment (yammesicka) [inline README.md]: Space `: `

File: README.md (hunk 9 above)
Comment (yammesicka) [reply README.md]: `"vscode": {`

File: example/example_agamm/.comeback [hunk 11]
```diff
@@ -1,3 +1,7 @@
-pycharm:
-  cwd: D:\Development\comeback
-
+[
+   {
+      "pycharm":{
```
Comment (yammesicka) [inline example/example_agamm/.comeback]: `: {`

File: example/example_agamm/.comeback [hunk 12]
```diff
@@ -1,3 +1,7 @@
...
+         "cwd":"D:\\Development\\comeback"
+      }
+   }
+]
```
Comment (yammesicka) [inline example/example_agamm/.comeback]: Newline

File: example/example_agamm/.comeback [hunk 13]
```diff
@@ -1,3 +1,7 @@
-pycharm:
-  cwd: D:\Development\comeback
-
+[
+   {
+      "pycharm":{
+         "cwd":"D:\\Development\\comeback"
```
Comment (yammesicka) [inline example/example_agamm/.comeback]: `: `

File: example/example_basic/.comeback [hunk 14]
```diff
@@ -1,4 +1,12 @@
-vscode: 
-  cwd: ~/dev/myproject
-chrome:
-  url: http://localhost:8080/
\ No newline at end of file
+[  
+   {  
+      "vscode":{
```
Comment (yammesicka) [inline example/example_basic/.comeback]: `: {`

File: example/example_basic/.comeback [hunk 15]
```diff
@@ -1,4 +1,12 @@
-vscode: 
-  cwd: ~/dev/myproject
-chrome:
-  url: http://localhost:8080/
\ No newline at end of file
+[  
+   {  
+      "vscode":{  
+         "cwd":"~/dev/myproject"
```
Comment (yammesicka) [inline example/example_basic/.comeback]: `: `

File: comeback/recipe_manager.py (hunk 4 above)
Comment (agamm) [reply comeback/recipe_manager.py]: Renamed to create_file

File: comeback/plugins/pycharm/main.py [hunk 16]
```diff
@@ -68,7 +68,7 @@ def run_mac(cwd: str) -> utils.RUN_STATUS:
         return False, "Can't find PyCharm in Applications, did you install it?"
 
     pycharm_path = str(results[0])
-    utils.run(f'open "{pycharm_path}" --args {cwd}', use_shell=True)
+    utils.run(f'open "{pycharm_path}" --args {cwd}', use_shell=False)
```
Comment (yammesicka) [inline comeback/plugins/pycharm/main.py:71]: 🎉

File: comeback/main.py [hunk 17]
```diff
@@ -41,70 +34,8 @@ def parse_args(args: Optional[str] = None) -> Dict[str, str]:
...
-    except IOError as e:
-        click.echo(f'Could not read file {config_path} because {e}')
-    except yaml.YAMLError as exc:
-        click.echo('YAML Error: ' + str(exc))
-    return None
-
-
 def get_config_path() -> pathlib.Path:
-    cwd_path = paths.CURRENT_DIR / '.comeback'
+    cwd_path = paths.CURRENT_DIR / recipe.RECIPE_FILENAME
```
Comment (yammesicka) [inline comeback/main.py]: Change to `FILENAME` or `DEFAULT_FILENAME`

File: comeback/main.py [hunk 18]
```diff
@@ -186,15 +91,14 @@ def choose_last_used() -> None:
...
-    IS_VERBOSE = verbose
+    utils.IS_VERBOSE = verbose
 
     if ctx.invoked_subcommand is not None:
         verbose_echo('Invoking subcommand: %s' % ctx.invoked_subcommand)
         return
 
     if init:
-        create_comeback_file_here()
+        recipe.create_file()
```
Comment (yammesicka) [inline comeback/main.py]: `recipe.create`?

File: comeback/main.py [hunk 19]
```diff
@@ -209,4 +113,9 @@ def cli(ctx: click.Context, init: bool, verbose: bool, last_used: bool) \
 @click.argument('plugin_params', required=False)
 def run(plugin: str, plugin_params: str) -> None:
     verbose_echo(f'Running plugins: {plugin} with args: {plugin_params}')
-    load_plugin(plugin, parse_args(plugin_params))
+    module = manager.load(plugin)
```
Comment (yammesicka) [inline comeback/main.py]: Maybe here it'll be better to use the full naming: `plugins.manager.load`

File: comeback/plugins/manager.py [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+
+
+def call(module: ModuleType,
+         **plugin_params: Dict[str, Any]) -> None:
+    success = False
+    err = None
+    try:
+        success, err = module.run_plugin(**plugin_params)
+    except TypeError as e:
+        click.echo(
```
Comment (yammesicka) [inline comeback/plugins/manager.py]: `success, err = False, ...` Instead?

File: comeback/plugins/manager.py [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+            f'There was a problem executing the plugin {module.__name__}: {e}')
+        exit()
+
+    if not success:
+        click.echo(
+            f'There was a problem executing the plugin {module.__name__}: {err}')
+        exit()
```
Comment (yammesicka) [inline comeback/plugins/manager.py]: exit with bad exit code

File: comeback/recipe.py [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+        with open(recipe_path, 'r') as fd:
+            return json.load(fd)
+    except IOError as e:
+        click.echo(f'Could not read file {recipe_path} because {e}')
+    except json.JSONDecodeError as exc:
+        click.echo('JSON Error: ' + str(exc))
+    return None
+
+
+def run(recipe: List[Dict[str, Any]]) -> None:
```
Comment (yammesicka) [inline comeback/recipe.py]: Function is too complicated. Perhaps we should split it and make it more readable

File: comeback/recipe.py [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+    for plugin_dict in recipe:
+        for plugin_name, plugin_params in plugin_dict.items():
+            if not manager.does_exists(plugin_name):
+                exit()
+
+            verbose_echo(f'Starting {plugin_name}...')
+            verbose_echo(f'\tParams {plugin_params}...')
+
+            plugin = manager.load(plugin_name)
```
Comment (yammesicka) [inline comeback/recipe.py]: Use `plugin.manager`

File: comeback/recipe.py (hunk 22 above)
Comment (agamm) [reply comeback/recipe.py]: Please suggest something more readable :)

File: comeback/plugins/pycharm/main.py (hunk 16 above)
Comment (yammesicka) [inline comeback/plugins/pycharm/main.py:71]: Use '

File: comeback/plugins/pycharm/main.py [hunk 24]
```diff
@@ -68,7 +68,7 @@ def run_mac(cwd: str) -> utils.RUN_STATUS:
         return False, "Can't find PyCharm in Applications, did you install it?"
```
Comment (yammesicka) [inline comeback/plugins/pycharm/main.py:68]: Use '

File: comeback/recipe.py [hunk 25]
```diff
@@ -0,0 +1,73 @@
+import click
+import json
+import pathlib
+from typing import Any, Dict, Optional, List
+
+from comeback import config, plugin_manager
+from comeback import paths
+from comeback.utils import verbose_echo
+
+FILENAME = ".comeback"
```
Comment (yammesicka) [inline comeback/recipe.py]: Use '

File: comeback/plugins/pycharm/main.py (hunk 16 above)
Comment (agamm) [reply comeback/plugins/pycharm/main.py:71]: run...

File: comeback/plugin_manager.py [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+    success = False
+    err = None
+    try:
+        success, err = module.run_plugin(**plugin_params)
+    except TypeError as e:
+        success, err = False, str(e)
+
+    if not success:
+        click.echo(
+            f'There was a problem executing the plugin {module.__name__}: {err}')
```
Comment (yammesicka) [inline comeback/plugin_manager.py]: Line too long

File: comeback/plugin_manager.py [hunk 27]
```diff
@@ -0,0 +1,43 @@
+import click
```
Comment (yammesicka) [inline comeback/plugin_manager.py]: Imports in wrong order

**Comment by agamm** [comment]
Can I bribe you to make this pass? @yammesicka

**Review by yammesicka — APPROVED** [review]

File: comeback/main.py [hunk 28]
```diff
@@ -1,25 +1,17 @@
-import importlib
+import click
```
Comment (yammesicka) [inline comeback/main.py]: import order

File: comeback/paths.py [hunk 29]
```diff
@@ -2,10 +2,13 @@
 import pathlib
 from typing import NamedTuple
 
+from comeback.recipe import FILENAME
```
Comment (yammesicka) [inline comeback/paths.py]: `from comeback import recipe`

**Pull request merged**